### ISSUE TITLE
2.7 - ongoing list of incompatibilities w/ PHP 7

### ISSUE THREAD

**Issue by sebastienbarre** [cakephp/cakephp#7671]
Per tangent in #5602, here is a list of (possible) ongoing incompatibilities between PHP 7 and the 2.7 branch. It would be great if 2.x could handle PHP 7.
Feel free to pick one and create a separate issue (?)

### Parse error are no longer reported correctly
Per [PHP: Backward incompatible changes - Manual](http://php.net/manual/en/migration70.incompatible.php), "Many fatal and recoverable fatal errors have been converted to exceptions in PHP 7. These error exceptions inherit from the Error class, which itself implements the Throwable interface (the new base interface all exceptions inherit). [...] Parser errors now throw a ParseError object".

So as soon as a syntax error is encountered, the following happens, which isn't quite useful:
```
Uncaught TypeError: Argument 1 passed to ErrorHandler::handleException() must be an instance of Exception, instance of ParseError given in lib\Cake\Error\ErrorHandler.php:116 Stack trace: #0 [internal function]: ErrorHandler::handleException(Object(ParseError)) #1 {main} thrown
File: lib\Cake\Error\ErrorHandler.php
Line: 116
```
My gut feeling here is `set_exception_handler` routes *both* the `Exception` and `Error` classes that implements the new `Throwable`, but since the signature of `handleException` only accepts an `Exception`, it breaks. Updating the signature to accept `Throwable` would work, but this is new to PHP 7. Looking at `handleException` [in 3.0](https://github.com/cakephp/cakephp/blob/master/src/Error/BaseErrorHandler.php#L153), I see the signature is still the same, so how did you guy solve it?

**Comment by tanuck** [comment]
Something that might be of interest... The Whoops library came across the same problems when trying to get their test suite passing on PHP7. There's an extensive discussion on it here: https://github.com/filp/whoops/pull/284

**Comment by markstory** [comment]
I'm open to removing the typehint on the existing methods. I would rather have that kind of change than not working in PHP7.

**Comment by dereuromark** [comment]
We could do all that for 2.8 - including travis support for PHP7.
This would give us the possibilities for further changes that might have to be communicated more clearly.
It will not take too long for 2.8 to be out anyway, so that could be the version fully supporting PHP7 then IMO.

**Comment by dereuromark** [comment]
I started a 2.8 branch with the proposed fixes:
https://github.com/cakephp/cakephp/compare/2.8...2.8-PHP7

Feel free to append your commits on top.

I wonder if we also have to change other places like

   } catch (Exception $e) {

inside the codebase.

**Comment by markstory** [comment]
@dereuromark Making 2.8 compatible with PHP7 is a great idea.

**Comment by tanuck** [comment]
@dereuromark could you use `Throwable` in the docblock to be more succinct?

**Comment by jhli** [comment]
Personally I think only if php7 adoption will be very quick there is a need to adopt php7 compatibility now. Rather php 5.6 (or 5.7) will be the new 5.2/5.3 living for ages. I am not against the changes but removimg typehints may make a difference for some.

**Comment by sebastienbarre** [comment]
Making 2.8 compatible with PHP7 would be indeed great. My gut feeling is that it should be done sooner rather than later, if only to remain competitive with other frameworks that are already compatible with PHP7. I can't predict the future but the hype around PHP7 performance is real, and I would not be surprised if it was adopted quickly. Whereas I had seen minor improvements switching from 5.5 to 5.6, then 5.6 to 5.7 (say, in the order of 10% faster), the boost provided by PHP7 is at a whole different level. I've measured our codebase running almost twice as fast in the vast majority of our test cases. This is nothing to sneeze at.

**Comment by ravage84** [comment]
A minor version, such as 2.8, is a fitting place for such changes.
But we do not need to rush anything, if there will be a version 2.9.

In the end, compatibility for CakePHP 2.x with PHP 7 would extend the lifespan of the 2.x channel for years. Question is, do we actually want that?

**Comment by dereuromark** [comment]
I wouldnt be so sure about that. I think 2.8 should really be it, as many people will soon already start to actually work with PHP7 as default. Months maybe, not years.

**Comment by ionas** [comment]
I have no problem with PHP 7 upgrades under one circumstance: **CakePHP 2.8/2.9 should NOT add non-BC breaks just because it is required for PHP7**.

**Why?**
There are MANY more people out simply maintaining 2.x applications than writing new ones. The less they have to change the better.

*Why do they have to upgrade within the 2.x tree anyway?*
Because the last 2.x stable release will be, if at all, the one maintained and supplied with bug fixes and security fixes (like 1.3.x before). So it is more or less mandatory to upgrade from 2.5/2.6/2.7 to 2.8 or 2.9. This upgrade should be hassle free. If you start new applications and want new features, there is CakePHP 3.x already.

Thus my point of view:
- If you want to maintain 2.x thats really great for the community, the image of CakePHP by having LTS
- Maintaining the 2.x tree should introduce AS LITTLE BC changes at this point as possible.

As for 3.x I am on the other side. Some non-BC changes are good, going for a new version name (4.x) faster than with 2.x -> 3.x for signaling bigger BC breaks is fine, too. Adopting PHP 7 and by that being required to change some things that are NON-BC in 3.x is also no problem. If you want recent powerful features, you gotta adapt to more speedy upgrade cycles. For 2.x I don't see the reason to pursuit that way; if it hurts more than it helps, it might not be good.

Final Note: To all those maintaining 2.x - especially @dereuromark : I love you guys. Thanks for all the hard work. It is really appreciated.

**Comment by AD7six** [comment]
> CakePHP 2.8/2.9 should NOT add non-BC breaks

Double negatives are hard to read:

> CakePHP 2.8/2.9 should add BC breaks

Assuming you mean should _not_ add bc breaks - Removing the type hint of the error handler isn't going to prevent it working with php 5.x.

**Comment by dereuromark** [comment]
@ionas No one will take PHP5 support away in 2.x - I wonder how/where you would get those ideas. "Adding PHP7 support" implies a very clear "on top" IMO.

**Comment by ADmad** [comment]
I believe we also need to address the issue being able to run test suite of 2.x on PHP 7.

**Comment by ionas** [comment]
> CakePHP 2.8/2.9 should add BC breaks

My stance again ;):
- CakePHP 2.8/2.9 should **try** to do *only backwards-compatible changes* as there are many projects that need maintenance and will require the last stable 2.x version to do that.
- CakePHP 2.8/2.9 should not add backwards-compatibility-breaking changes just because of a possible PHP7 adoption.

The alternative stance to that IMHO is that 2.8/2.9 should make sure it runs on PHP7 VERY quickly so that all further upgrades/maintenance issues are clear up the road for those simply maintaining 2.x apps (the dominant 2.x usage).

**Comment by sebastienbarre** [comment]
Just my 2 cents, since I opened that thread. I was not expecting (or requesting, god forbid) any new fancy PHP7-only features to show up in 2.x. With PHP7 out in two days, I was trying to push for the smallest set of changes that would ensure the test suite would pass in PHP7, and bring 2.x users the confidence that we can upgrade to PHP7 and benefit from that big performance boost.

As a side note, I really hope 2.x will be "around" for years, not months. As CakePHP developers and maintainers I can imagine you are proud of 3.x, as you should be, but we can't all afford to upgrade to 3.x, especially with the new ORM -- this would be very time consuming and costly for us. Upgrading from 1.x was a beast already, it's not a huge app but we are talking 115K+ lines of PHP, 45K+ lines of CTP, 11K+ lines of JavaScript, etc., and we can't have any downtime.

Thanks

**Comment by dereuromark** [comment]
@sebastienbarre That is exactly how everyone (expect maybe @ionas ) understood you and this issue :)
And what I started aims to do exactly that, feel free to finalize my work.

**Comment by markstory** [comment]
I too have no interest in adding breaking changes to 2.x. If there are code paths that do not work with php7 I would rather add new code paths that people using php7 and 2.x can use. We have already done that kind of thing with String->CakeText rename for example.

@sebastienbarre 2.x will enjoy the same long life that 1.x did from a maintenance release/support point of view.

**Comment by dereuromark** [comment]
PR is open, please contribute there.

**Comment by mayankparek** [comment]
My project in cakephp 2.7.11 and i used ubuntu 16.04 . when i execute project it gives error like this
Error: Uncaught TypeError: Argument 1 passed to ErrorHandler::handleException() must be an instance of Exception, instance of ParseError given in /var/www/html/cake/lib/Cake/Error/ErrorHandler.php:116 Stack trace: #0 [internal function]: ErrorHandler::handleException(Object(ParseError)) #1 {main} thrown	
File: /var/www/html/cake/lib/Cake/Error/ErrorHandler.php	
Line: 116

and i dont want to use php 7.0 then is it possible ?

**Comment by dereuromark** [comment]
As you can see this was all merged in to 2.8.
Also, the docs and migration guides clearly point out that in 2.x it is `>= 2.8` which support PHP7+.
